File: app/src/main/java/work/petrichor/petrichor/modules/CameraInterface.java
package work.petrichor.petrichor.modules;

import android.graphics.ImageFormat;
import android.hardware.Camera;
import android.hardware.Camera.*;
import android.view.SurfaceHolder;

import java.io.IOException;
import java.util.List;
import java.util.concurrent.ExecutionException;

import work.petrichor.seeta.FaceInfo;
import work.petrichor.seeta.ImageData;

/**
 * @author Mitscherlich.
 * @date 2017/7/26.
 **/

public class CameraInterface {

    private Camera mCamera;
    private Parameters mParams;
    private FaceDetectionTask mTask;
    private String modelPath = "";

    private boolean isPreview = false;

    private static CameraInterface instance;

    public static final int ON_SEETA_FACE_DETECTION_MESSAGE = 100;
    public static final int ON_GOOGLE_FACE_DETECTION_MESSAGE = 101;

    public static synchronized CameraInterface getInstance() {
        if (instance == null) instance = new CameraInterface();
        return instance;
    }

    private CameraInterface() {}

    public void openCamera(CameraOpenCallback callback) {
        mCamera = Camera.open(getFrontCameraID());
        callback.hasCameraOpened();
    }

    public void initParameters() {
        mParams = mCamera.getParameters();

        // 设置预览参数
        Size previewSize = getPreviewSize(mParams);
        mParams.setPreviewSize(previewSize.width / 3, previewSize.height / 3);
        mParams.setPreviewFormat(ImageFormat.YV12);

        // 设置拍照参数
        Size pictureSize = getPictureSize(mParams);
        mParams.setPictureSize(pictureSize.width, pictureSize.height);

        // 设置自动对焦模式
        String focusMode = getFocusMode(mParams);
        mParams.setFocusMode(focusMode);

        mCamera.setParameters(mParams);
        if (previewSize.width > previewSize.height)
            mCamera.setDisplayOrientation(90);
        else
            mCamera.setDisplayOrientation(0);
    }

    public void setPreviewDisplay(SurfaceHolder holder) {
        try {
            mCamera.setPreviewDisplay(holder);
        } catch (IOException e) {
            e.printStackTrace();
        }
    }

    public void setFaceDetectionListener(Camera.FaceDetectionListener listener) {
        mCamera.setFaceDetectionListener(listener);
    }

    public void setFaceDetectionTask(String modelPath) {
        this.modelPath = modelPath;
        mTask = new FaceDetectionTask(modelPath);
    }

    public void setPreviewListener(final FaceDetectionListener listener) {
        mCamera.setPreviewCallback(new PreviewCallback() {
            @Override
            public void onPreviewFrame(byte[] bytes, Camera camera) {
                switch (mTask.getStatus()) {
                    case RUNNING:
                        return;
                    case PENDING:
                        mTask.cancel(false);
                }

                mTask = new FaceDetectionTask(modelPath);

                int width = mParams.getPreviewSize().width;
                int height = mParams.getPreviewSize().height;
                int channels = bytes.length / (width * height);

                if (bytes.length > 0) {
//                    long now = System.currentTimeMillis();
//                    byte[] data = filipFrontCameraPreviewData(bytes, width, height);
//                    long ms = System.currentTimeMillis() - now;
//                    android.util.Log.i("CameraInterface", String.format("filip front camera data in %dms", ms));
//                    ImageData frame = new ImageData(data, width, height, channels);
                    ImageData frame = new ImageData(bytes, width, height, channels);
                    try {
                        List<FaceInfo> faces = mTask.execute(frame).get();
                        if (faces == null) return;
                        else listener.onFaceDetection(faces);
                    } catch (InterruptedException | ExecutionException e) {
                        e.printStackTrace();
                    }
                }
            }
        });
    }

    public void startPreview() {
        if (isPreview)
            stopPreview();

        mCamera.startPreview();
        isPreview = true;

        if (mCamera.getParameters().getMaxNumDetectedFaces() > 0)
            mCamera.startFaceDetection();
    }

    public void stopPreview() {
        try {
            if (mCamera != null) {
                if (mCamera.getParameters().getMaxNumDetectedFaces() > 0)
                    mCamera.stopFaceDetection();

                mCamera.setPreviewCallback(null);
                mCamera.setPreviewDisplay(null);
                mCamera.stopPreview();
                mCamera.release();
                mCamera = null;
                isPreview = false;
            }
        } catch (IOException e) {
            e.printStackTrace();
        }
    }

    private int getFrontCameraID() {
//        return Camera.getNumberOfCameras() > 1 ? 1 : 0;
        return 0;
    }

    private Size getPreviewSize(Parameters params) {
        List<Size> sizes = params.getSupportedPreviewSizes();
        return sizes.get(0);
    }

    private String getFocusMode(Parameters params) {
        List<String> modes = params.getSupportedFocusModes();
        if (modes.contains("continuous-video"))
            return Parameters.FOCUS_MODE_CONTINUOUS_VIDEO;
        else
            return Parameters.FOCUS_MODE_CONTINUOUS_PICTURE;
    }

    private Size getPictureSize(Parameters params) {
        List<Size> sizes = params.getSupportedPictureSizes();
        return sizes.get(0);
    }

//    private byte[] filipFrontCameraPreviewData(byte[] src, int cols, int rows) {
//        byte[] dst = new byte[src.length];
//
//        for (int i = 0; i < rows; ++i)
//            for (int j = 0; j < cols; ++j)
//                dst[(rows - i - 1) * cols + j] = src[cols * i + j];
//
//        return dst;
//    }

    public interface CameraOpenCallback {
        void hasCameraOpened();
    }

    public interface FaceDetectionListener {
        void onFaceDetection(List<FaceInfo> faces);
    }
}

File: app/src/main/java/work/petrichor/petrichor/modules/CameraInterface.kt
package work.petrichor.petrichor.modules

import android.graphics.ImageFormat
import android.hardware.Camera
import android.hardware.Camera.*
import android.os.AsyncTask
import android.view.SurfaceHolder
import work.petrichor.facedetection.format.FaceInfo
import work.petrichor.facedetection.format.ImageData
import java.util.concurrent.LinkedBlockingDeque

/**
 * @author  Mitscherlich.
 * @data    2017/7/24.
 */
object CameraInterface {

    val ON_SEETA_FACE_DETECTION_MESSAGE = 100
    val ON_GOOGLE_FACE_DETECTION_MESSAGE = 101

    private var mCamera: Camera? = null
    private var isPreview: Boolean = false
    private var modelPath: String = ""

    private lateinit var mTask: FaceDetectionTask

    private lateinit var mParams: Parameters
    private lateinit var faceDetectionQueue: FaceDetectionQueue

    private var frameQueue: LinkedBlockingDeque<ImageData> = LinkedBlockingDeque()
    fun openCamera(callback: CameraOpenCallback) {
        mCamera = Camera.open(getFrontCamera())
        callback.hasCameraOpened()
    }

    fun initParameters() {
        mParams = mCamera!!.parameters

        val pictureSize = getPictureSize(mParams)
        mParams.setPictureSize(pictureSize.width, pictureSize.height)

        val previewSize = getPreviewSize(mParams)
        mParams.setPreviewSize(previewSize.width, previewSize.height)
        mParams.previewFormat = ImageFormat.YV12

        mParams.focusMode = getFocusMode(mParams)

        mCamera!!.parameters = mParams
        mCamera!!.setDisplayOrientation(90)
    }

    /**
     * 设置预览视图句柄
     * @param holder 从 CameraView 获得的绘图句柄
     * **/
    fun setPreviewDisplay(holder: SurfaceHolder) {
        mCamera!!.setPreviewDisplay(holder)
    }

    /**
     * 设置 Seetaface 人脸检测监听对象
     * @param modelPath
     * @param listener
     * **/
    fun setDetectionListener(modelPath: String, listener: FaceDetectionListener) {
        faceDetectionQueue = FaceDetectionQueue(frameQueue, modelPath, listener)
//        this.modelPath = modelPath
//        mTask = FaceDetectionTask(modelPath)
        mCamera!!.setPreviewCallback({ data, _ ->
            if (data == null || data.isEmpty()) return@setPreviewCallback
            val frame = ImageData(data, mParams.previewSize.width, mParams.previewSize.height, ImageData.FRAME_FORMAT_YV12)
            frameQueue.put(frame)
//            if (mTask.status == AsyncTask.Status.RUNNING) return@setPreviewCallback
//            if (mTask.status == AsyncTask.Status.PENDING) mTask.cancel(false)
//            mTask = FaceDetectionTask(CameraInterface.modelPath)
//            val faces = mTask.execute(ImageData(data, mParams.previewSize.width, mParams.previewSize.height, ImageData.FRAME_FORMAT_YV12)).get()
//            listener.onFaceDetection(faces)
        })
    }

    /**
     * 设置 Google 人脸检测监听对象
     * @param listener
     * **/
    fun setDetectionListener(listener: Camera.FaceDetectionListener) {
        mCamera!!.setFaceDetectionListener(listener)
    }

    /**
     * 开始预览
     * **/
    fun startPreview() {
        if (isPreview)
            mCamera!!.stopPreview()

        mCamera!!.startPreview()
        isPreview = true

        // 先判断设备是否支持人脸检测
        if (mParams.maxNumDetectedFaces > 0)
            mCamera!!.startFaceDetection()
    }

    /**
     * 停止预览
     * **/
    fun stopPreview() {
        if (mCamera != null) {
            if (mParams.maxNumDetectedFaces > 0) {
                mCamera!!.stopFaceDetection()
                mCamera!!.setFaceDetectionListener(null)
            }

            mCamera!!.setPreviewCallback(null)
            mCamera!!.setPreviewDisplay(null)
            mCamera!!.stopPreview()
            mCamera!!.release()
            mCamera = null
            isPreview = false
        }
    }

    /**
     * 获取中间支持的图像 Size
     * @param params 相机参数
     * @return 中值拍照 Size
     * **/
    private fun getPictureSize(params: Parameters): Size {
        val pictureSize = params.supportedPictureSizes
        return pictureSize[0]
    }

    /**
     * 获取中间支持的预览 Size
     * @param params 相机参数
     * @return 中值预览 Size
     * **/
    private fun getPreviewSize(params: Parameters): Size {
        val previewSize = params.supportedPreviewSizes
        return previewSize[0]
    }

    /**
     * 获取自动对焦模式
     * @param params 相机参数
     * @return 预览性质
     * **/
    private fun getFocusMode(params: Parameters): String {
        val focusModes = params.supportedFocusModes

        return if (focusModes.contains("continuous-video"))
            Parameters.FOCUS_MODE_CONTINUOUS_VIDEO
        else
            Parameters.FOCUS_MODE_CONTINUOUS_PICTURE
    }

    private fun getFrontCamera(): Int = if (Camera.getNumberOfCameras() > 1) 1 else 0

    interface CameraOpenCallback {
        fun hasCameraOpened()
    }

    interface FaceDetectionListener {
        fun onFaceDetection(faces: Array<FaceInfo>)
    }
}